lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -311,95 +298,49 @@ define((require, exports, module) ->
-          }
-
-          # after finding boundary, tokenIterator is exactly on the last token
-          # of ending sequence, so we step back until we're on the last token
-          # to determine inner boundary
-          for i in [0...finishedSequence.length-1]
-            moveFromBoundary()
-
-          { row: curTokenRow, column: curTokenColumn } = tokenIterator.getCurrentTokenPosition()
-          curTokenLength = tokenIterator.getCurrentToken().value.length
-
-          equationInnerBoundary = {
-            row: curTokenRow
-            column: curTokenColumn + (if start then curTokenLength else 0)
-          }
-
-          # to move inside the sequence, we move token iterator one more time
-          moveFromBoundary()
-
-          return {
-            id: maybeFinishedSequenceId
-            outer: equationOuterBoundary
-            inner: equationInnerBoundary
-          }
+      currentToken = tokenIterator.getCurrentToken()
+      { row: prevRow } = tokenIterator.getCurrentTokenPosition()
+      # TODO: magic string? importing is hard though
+      boundaryCorrect = true
+      while LatexParsingContext.isType(currentToken, "equation")
+        currentToken = moveToBoundary()
+        if not currentToken?
+          boundaryCorrect = false
+          break
+        { row: currentRow } = tokenIterator.getCurrentTokenPosition()
+        # Empty string always means that equation state is popped from state stack.
+        # Unfortunately, empty string is not tokenized at all, and TokenIterator
+        # just skips it altogether, so we have to handle this manually here.
+        if Math.abs(currentRow - prevRow) > 1
+          boundaryCorrect = false
+          break
+        prevRow = currentRow
+
+      if currentToken? and LatexParsingContext.isType(currentToken, "error")
+        boundaryCorrect = false
+
+      moveFromBoundary()
+
+      { row: curTokenRow, column: curTokenColumn } = tokenIterator.getCurrentTokenPosition()
+      curTokenLength = tokenIterator.getCurrentToken().value.length
+      return {
+        correct: boundaryCorrect
+        row: curTokenRow
+        column: curTokenColumn + if start then 0 else curTokenLength
+      }
 
     getEquationRange: (row, column) ->
       tokenIterator = new TokenIterator(@editor.getSession(), row, column)
-      end = @getBoundary(tokenIterator, false)
       start = @getBoundary(tokenIterator, true)
+      tokenIterator = new TokenIterator(@editor.getSession(), row, column)
+      end = @getBoundary(tokenIterator, false)
 
-      if not (start? and end?)
-        return {
-          outer: null
-          inner: null
-          start: null
-          end: null
-        }
-
-      { id: endId, outer: outerEnd, inner: innerEnd } = end
-      { id: startId, outer: outerStart, inner: innerStart } = start
+      { correct: startCorrect, row: startRow, column: startColumn } = start
+      { correct: endCorrect,   row: endRow,   column: endColumn } = end
```

Comment:
ditto, what's the purpose of destructuring here? Why don't you just write `start.correct and end.correct` in the expression below?